### PR TITLE
Refactor PageLoader with Flow

### PR DESCRIPTION
**Pull request by revonateB0T**

### CODE REVIEW

File: app/src/main/kotlin/com/hippo/ehviewer/download/DownloadManager.kt [hunk 1]
```diff
@@ -597,98 +599,82 @@ object DownloadManager : OnSpiderListener, CoroutineScope {
...
-        receivedSize: Long,
-        bytesRead: Int,
+        receivedSize: Flow<Long>,
     ) {
         launch {
-            mSpeedReminder.onDownload(index, contentLength, receivedSize, bytesRead)
+            receivedSize.runningFold(0L) { prev, curr ->
+                mSpeedReminder.onDownload(index, contentLength, curr, (curr - prev).toInt())
+                curr
+            }.collect()
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/download/DownloadManager.kt:623]: nit: use `launchIn` instead.

File: app/src/main/kotlin/com/hippo/ehviewer/gallery/EhPageLoader.kt [hunk 2]
```diff
@@ -21,87 +21,68 @@ import com.hippo.ehviewer.image.Image
...
         super.start()
-        if (!::mSpiderQueen.isInitialized) {
-            mSpiderQueen = obtainSpiderQueen(mGalleryInfo, SpiderQueen.MODE_READ)
-            mSpiderQueen.addOnSpiderListener(this)
+        if (!::spiderQueen.isInitialized) {
+            spiderQueen = obtainSpiderQueen(mGalleryInfo, SpiderQueen.MODE_READ)
+            sourceFlow = callbackFlow {
+                val listener = object : SpiderQueen.OnSpiderListener {
+                    override suspend fun onPageDownload(index: Int, contentLength: Long, receivedSize: Flow<Long>) {
+                        send(PageEvent.Progress(index, receivedSize.map { it.toFloat() / contentLength }))
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/gallery/EhPageLoader.kt:42]: nit: check if `contentLength > 0`.

File: app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderQueen.kt [hunk 3]
```diff
@@ -455,14 +456,14 @@ class SpiderQueen private constructor(val galleryInfo: GalleryInfo) : CoroutineS
...
-        fun onGet509(index: Int)
-        fun onPageDownload(index: Int, contentLength: Long, receivedSize: Long, bytesRead: Int)
-        fun onPageSuccess(index: Int, finished: Int, downloaded: Int, total: Int)
-        fun onPageFailure(index: Int, error: String?, finished: Int, downloaded: Int, total: Int)
-        fun onFinish(finished: Int, downloaded: Int, total: Int)
-        fun onGetImageSuccess(index: Int, image: Image?)
-        fun onGetImageFailure(index: Int, error: String?)
+        suspend fun onGetPages(pages: Int) {}
+        suspend fun onGet509(index: Int) {}
+        suspend fun onPageDownload(index: Int, contentLength: Long, progressFlow: Flow<Long>) {}
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderQueen.kt:461]: The name `progressFlow` is less descriptive.

File: app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderQueen.kt [hunk 4]
```diff
@@ -455,14 +456,14 @@ class SpiderQueen private constructor(val galleryInfo: GalleryInfo) : CoroutineS
...
+        suspend fun onPageSuccess(index: Int, finished: Int, downloaded: Int, total: Int) {}
+        suspend fun onPageFailure(index: Int, error: String?, finished: Int, downloaded: Int, total: Int) {}
+        suspend fun onFinish(finished: Int, downloaded: Int, total: Int) {}
+        suspend fun onGetImageSuccess(index: Int, image: Image?) {}
+        suspend fun onGetImageFailure(index: Int, error: String?) {}
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderQueen.kt:466]: Make `error` and `image` non-null.

File: app/src/main/kotlin/eu/kanade/tachiyomi/ui/reader/loader/PageLoader.kt [hunk 5]
```diff
@@ -37,9 +43,27 @@ abstract class PageLoader {
         )
     }
 
+    protected abstract val sourceFlow: Flow<PageEvent>
```
Comment (FooIbar) [inline app/src/main/kotlin/eu/kanade/tachiyomi/ui/reader/loader/PageLoader.kt:46]: The name is not very clear.

File: app/src/main/kotlin/eu/kanade/tachiyomi/ui/reader/loader/PageLoader.kt [hunk 6]
```diff
@@ -37,9 +43,27 @@ abstract class PageLoader {
...
+                    is PageEvent.Error -> PageStatus.Error(event.error)
+                    is PageEvent.Progress -> PageStatus.Loading(event.progress.stateIn(this, SharingStarted.Eagerly, 0f))
+                    is PageEvent.Success -> event.image.let {
+                        cache.put(index, it)
+                        if (it.hasQrCode) PageStatus.Blocked(it) else PageStatus.Ready(it)
+                    }
+                    is PageEvent.Wait -> PageStatus.Queued
+                }
+            }
+            val overwrite = broadcast.filter { (i, _) -> index == i }.map { (_, e) -> e }
```
Comment (FooIbar) [inline app/src/main/kotlin/eu/kanade/tachiyomi/ui/reader/loader/PageLoader.kt:64]: nit: `mapNotNull`

File: app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderQueen.kt (hunk 4 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderQueen.kt:466]: `error` cannot be non-null unless we refactor `SpiderQueen` later.

File: app/src/main/kotlin/com/hippo/ehviewer/gallery/EhPageLoader.kt (hunk 2 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/gallery/EhPageLoader.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderQueen.kt (hunk 3 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderQueen.kt]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderQueen.kt (hunk 4 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderQueen.kt]: Done

File: app/src/main/kotlin/eu/kanade/tachiyomi/ui/reader/loader/PageLoader.kt (hunk 6 above)
Comment (revonateB0T) [reply app/src/main/kotlin/eu/kanade/tachiyomi/ui/reader/loader/PageLoader.kt]: Done

**Pull request closed without merging**